packages/react-menu/src/components/MenuTrigger/useTriggerElement.ts

```diff
@@ -0,0 +1,85 @@
+import * as React from 'react';
+import { useMergedRefs } from '@fluentui/react-utilities';
+import { useKeyboardNavigationState } from '@fluentui/react-focus-management';
+import { MenuTriggerState } from './MenuTrigger.types';
+import { useMenuContext } from '../../menuContext';
+
+// Helper type to select on parts of state the hook uses
+type UseTriggerElementState = Pick<MenuTriggerState, 'children'>;
+
+export const useTriggerElement = (state: UseTriggerElementState): MenuTriggerState => {
+  const triggerRef = useMenuContext(context => context.triggerRef);
+  const setOpen = useMenuContext(context => context.setOpen);
+  const open = useMenuContext(context => context.open);
+  const triggerId = useMenuContext(context => context.triggerId);
+  const on = useMenuContext(context => context.on);
+  const { isNavigatingWithKeyboard } = useKeyboardNavigationState();
+
+  const child = React.Children.only(state.children);
+
+  const triggerProps: Partial<React.HTMLAttributes<HTMLElement>> = {
+    'aria-haspopup': true,
+    'aria-expanded': open,
+    id: triggerId,
+    ...(child.props || {}),
+  };
+  /**
+   * Opens menu when focused ONLY with keyboard focus
+   */
+  if (on.includes('focus')) {
+    triggerProps.onFocus = (e: React.FocusEvent) => {
+      if (isNavigatingWithKeyboard?.()) {
+        setOpen(true);
+      }
+      child.props?.onFocus?.(e);
+    };
+  }
+
+  /**
+   * Opens the menu when clicked
+   */
+  if (on.includes('click')) {
+    triggerProps.onClick = (e: React.MouseEvent) => {
+      if (!on.includes('context')) {
+        setOpen(true);
+      }
+
+      child.props?.onClick?.(e);
+    };
+  }
```

Comment:
There is some background from Northstar about it, these callbacks should be stable. It a real ask that we had from feature teams. For example:

```tsx
const MyMemoizedComponent = React.memo(/* some */);

// ---

function AwesomeComponent() {
  return (
    <MenuTrigger>
      {/* 💣 all "triggerProps" are unstable by a reference */}
      {/* this will break memoization and will produce unexpected renders */}
      <MyMemoizedComponent />
    </MenuTrigger>
  );
}
```

_Suggestion:_ Let's use `useEventCallback`

```ts
const handleClick = useEventCallback(() => {
  if (on.includes('click')) {
    triggerProps.onClick = (e: React.MouseEvent) => {
      if (!on.includes('context')) {
        setOpen(true);
      }

      child.props?.onClick?.(e);
    };
  }
})

//

React.cloneElement(child, {
  ...child.props,
  onClick: handleClick,
  // etc.
  ref: useMergedRefs(child.props.ref, triggerRef),
});
```

Reply:
That's a good point about cloning, thx